virttest/libvirt_vm.py

```diff
@@ -2137,6 +2137,46 @@ def detach_interface(self, option="", ignore_status=False,
                                       ignore_status=ignore_status,
                                       debug=debug)
 
+    def attach_device(self, xml_file, options="", ensure_boot=False, debug=False):
+        """
+        Attach a device to VM.
+
+        :param xml_file: The device xml for attach.
+        :param options: The options for device attach, which could be '--live',
+                '--config', '--current', '--persistent' or any combinations.
+                Default value is "" for live hotplug.
+        :param ensure_boot: Make sure the device attach happens after vm has
+                booted. That is necessary when checking the function of
+                hotplugged device.
+        :param debug: Open or close debug information.
+        """
+        if (ensure_boot):
```

Comment:
You should extract this part in order to avoid code duplication, something like
```python
def _ensure_booted(ensure_booted):
    if ensure_booted:
        ...
    self.wait_for_login()
```
```python
def attach_device(...):
    ...
    self._ensure_booted(ensure_booted)
    return virsh.attach_device(...)

def detach_device(...):
    ...
    self._ensure_booted(ensure_booted)
    return virsh.detach_device(...)
```